### PR TITLE
Support taking de-duplication snapshots based on time

### PR DESCRIPTION
**Pull request by 315157973**
Master Issue: #8237

### Motivation
Currently we take de-duplication snapshots based on size. If the topic has relatively low traffic, the de-duplication cursor will not move. This can cause messages that are not able to be deleted based on the retention policy. We should add a policy to take de-duplication snapshots based on time.

### Modifications
add a monitor to take snapshot base on time

### Verifying this change
TopicDuplicationTest#testDuplicationSnapshot

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 1]
```diff
@@ -297,6 +298,8 @@ public BrokerService(PulsarService pulsar) throws Exception {
                 .newSingleThreadScheduledExecutor(new DefaultThreadFactory("consumed-Ledgers-monitor"));
         this.ledgerFullMonitor =
                 Executors.newSingleThreadScheduledExecutor(new DefaultThreadFactory("ledger-full-monitor"));
+        this.deduplicationSnapshotMonitor =
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:301]: Can we save creating the pool if the feature is disabled

Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:301]: It looks like we are not shutting down the pool, leaking threads

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java [hunk 2]
```diff
@@ -413,6 +413,18 @@
...
+        category = CATEGORY_POLICIES,
+        doc = "How often is the thread pool scheduled to check whether a snapshot needs to be taken.(disable with value 0)"
+    )
+    private int brokerDeduplicationSnapshotFrequencyInSeconds = 120;
+    @FieldContext(
+        category = CATEGORY_POLICIES,
+        doc = "If this time interval is exceeded, a snapshot will be taken."
+            + "It will run simultaneously with `brokerDeduplicationEntriesInterval`"
+    )
+    private Integer brokerDeduplicationSnapshotIntervalSeconds = 120;
```
Comment (eolivelli) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:426]: Why Integer and not int?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java (hunk 2 above)
Comment (315157973) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:426]: Later, we need to support the namespace-level and topic-level. When it is null, it means that this level is not set. Therefore the packaging type is used.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 1 above)
Comment (315157973) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:301]: Oh Yes , I  always forget :sweat_smile:

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java (hunk 2 above)
Comment (eolivelli) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:426]: Makes sense.
Okay.
Thanks for your clarification

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**